frame/pools/src/lib.rs

```diff
@@ -0,0 +1,863 @@
+#[codec(mel_bound(T: Config))]
+#[scale_info(skip_type_params(T))]
+struct SubPoolsContainer<T: Config> {
+	/// A general, era agnostic pool of funds that have fully unbonded. The pools
+	/// of `self.with_era` will lazily be merged into into this pool if they are
+	/// older then `current_era - T::MAX_UNBONDING`.
+	no_era: UnbondPool<T>,
+	/// Map of era => unbond pools.
+	with_era: SubPoolsWithEra<T>,
+}
+
+impl<T: Config> SubPoolsContainer<T> {
+	/// Merge the oldest unbonding pool with an era into the general unbond pool with no associated
+	/// era.
+	fn maybe_merge_pools(mut self, current_era: EraIndex) -> Self {
+		if current_era < T::MaxUnbonding::get().into() {
+			// For the first `T::MAX_UNBONDING` eras of the chain we don't need to do anything.
+			// I.E. if `MAX_UNBONDING` is 5 and we are in era 4 we can add a pool for this era and
+			// have exactly `MAX_UNBONDING` pools.
+			return self
+		}
+
+		//  I.E. if `MAX_UNBONDING` is 5 and current era is 10, we only want to retain pools 6..=10.
+		let oldest_era_to_keep = current_era - T::MaxUnbonding::get().saturating_add(1);
+
+		let eras_to_remove: Vec<_> =
+			self.with_era.keys().cloned().filter(|era| *era < oldest_era_to_keep).collect();
+		for era in eras_to_remove {
+			if let Some(p) = self.with_era.remove(&era) {
+				self.no_era.points.saturating_add(p.points);
+				self.no_era.balance.saturating_add(p.balance);
+			} else {
+				// the world is broken
+			}
+		}
+
+		self
+	}
+}
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use frame_system::{ensure_signed, pallet_prelude::*};
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(crate) trait Store)]
+	#[pallet::generate_storage_info]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// The overarching event type.
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+		/// Weight information for extrinsics in this pallet.
+		// type WeightInfo: weights::WeightInfo;
+
+		/// The nominating balance.
+		type Currency: Currency<Self::AccountId>;
+
+		// Infallible method for converting `Currency::Balance` to `u128`.
+		type BalanceToU128: Convert<BalanceOf<Self>, u128>;
+
+		// Infallible method for converting `u128` to `Currency::Balance`.
+		type U128ToBalance: Convert<u128, BalanceOf<Self>>;
+
+		/// The interface for nominating.
+		type StakingInterface: StakingInterface<
+			Balance = BalanceOf<Self>,
+			AccountId = Self::AccountId,
+			LookupSource = <Self::Lookup as StaticLookup>::Source,
+		>;
+
+		/// The maximum amount of eras an unbonding pool can exist prior to being merged with the
+		/// `no_era	 pool. This should at least be greater then the `UnbondingDuration` for staking
+		/// so delegator have a chance to withdraw unbonded before their pool gets merged with the
+		/// `no_era` pool. This *must* at least be greater then the slash deffer duration.
+		type MaxUnbonding: Get<u32>;
```

Comment:
Does it make sense to make this be an associated type or function in staking interface?

Reply:
Yes, assuming staking interface is really only meant for pallet-pools

Reply:
even if not, it seems to me like that the `MaxUnbonding` is some information that will always be known and fulfilled by the implementor staking interface.